content/posts/wasi-experimental-http.md

```diff
@@ -0,0 +1,146 @@
+workloads that can be executed in WASI runtimes.
+
+Today, [we are releasing a library][gh] which adds experimental and _temporary_
+support for outgoing HTTP requests for WASI modules running in Wasmtime. It is
+important to note from the beginning that this library does not aim to replace
+the WASI sockets API, but rather provide a temporary workaround until the WASI
+networking API is stable, and we expect that once [the WASI sockets
+proposal][sockets-wip] gets adopted and implemented in language toolchains, the
+need for this library would go away.
+
+### Using the new HTTP library
+
+There are two aspects to a WebAssembly library - building guest Wasm modules,
+and adding runtime support for instantiating and executing those modules. Let's
+first see how to build a guest module in Rust:
+
+```rust
+use bytes::Bytes;
+use http;
+use wasi_experimental_http;
+
+#[no_mangle]
+pub extern "C" fn _start() {
+    let url = "https://<some-domain>/post".to_string();
+    let req = http::request::Builder::new()
+        .method(http::Method::POST)
+        .uri(&url)
+        .header("Content-Type", "text/plain")
+    let b = Bytes::from("sending a body in a POST request");
+    let req = req.body(Some(b)).unwrap();
+
+    let res = wasi_experimental_http::request(req).expect("cannot make request");
+    let str = std::str::from_utf8(&res.body()).unwrap().to_string();
+    // do something with the body
+}
+```
+
+From the start, the one thing to note is that we are reusing the Rust `Request`
+and `Response` structures from [the `http` crate][rust-http] (with `Bytes`
+request and response bodies), so this way of building HTTP requests should feel
+somewhat familiar to Rust developers. Then, a response can be sent and received
+using `wasi_experimental_http::request(req)`. This simple program can now be
+compiled to the `wasm32-wasi` target. AssemblyScript is another popular language
+that natively compiles to WebAssembly, and the library we are releasing also
+includes an AssemblyScript NPM package (with examples for how it can be used on
+[GitHub][gh]).
+
+In order to run the module in a WebAssembly runtime, we have to satisfy the new
+functionality, and we can do this in Wasmtime using the
+`wasi_experimental_http_wasmtime` crate:
+
+```rust
+use wasi_experimental_http_wasmtime;
+use wasmtime::*;
+use wasmtime_wasi::Wasi;
+
+let store = Store::default();
+let mut linker = Linker::new(&store);
+let wasi = Wasi::new(&store, ctx);
+
+// link the WASI core functions
+wasi.add_to_linker(&mut linker)?;
+
+// link the experimental HTTP support
+wasi_experimental_http_wasmtime::link_http(&mut linker, None)?;
+```
+
+The Wasmtime implementation also enables passing a list of allowed hosts - an
+optional and configurable list of domains or hosts that guest modules are
+allowed to send requests to. If a guest module attempts to send a request to a
+domain not explicitly allowed when the runtime was configured, it would receive
```

Comment:
"it receives" -- while the conditional is grammatically correct, it's hard to parse for a lot of readers so we typically use the simple present.